tests/rest/client/test_rooms.py

```diff
@@ -966,6 +970,81 @@ async def user_may_join_room(
         self.helper.join(self.room3, self.user2, expect_code=403, tok=self.tok2)
 
 
+class RoomAppserviceTsParamTestCase(RoomBase):
+    servlets = [
+        room.register_servlets,
+        synapse.rest.admin.register_servlets,
+        register.register_servlets,
+    ]
+
+    def prepare(self, reactor, clock, homeserver):
+        self.ts = 1
+
+        self.appservice_user, _ = self.register_appservice_user(
+            "as_user_potato", self.appservice.token
+        )
+
+        self.room = self.helper.create_room_as(
+            room_creator=self.appservice_user,
+            tok=self.appservice.token,
+            appservice_user_id=self.appservice_user,
+        )
+
+    def make_homeserver(self, reactor: MemoryReactor, clock: Clock) -> HomeServer:
+        config = self.default_config()
+
+        self.appservice = ApplicationService(
+            token="i_am_an_app_service",
+            hostname="test",
+            id="1234",
+            namespaces={"users": [{"regex": r"@as_user.*", "exclusive": True}]},
+            # Note: this user does not have to match the regex above
+            sender="@as_main:test",
+        )
+
+        mock_load_appservices = Mock(return_value=[self.appservice])
+        with patch(
+            "synapse.storage.databases.main.appservice.load_appservices",
+            mock_load_appservices,
+        ):
+            hs = self.setup_test_homeserver(config=config)
+        return hs
+
+    def test_state_event_msc3316_ts(self):
+        normal_user = self.register_user("thomas", "hackme")
+        event_id = self.helper.invite(
+            self.room,
+            self.appservice_user,
+            normal_user,
+            tok=self.appservice.token,
+            msc3316_ts=self.ts,
+            appservice_user_id=self.appservice_user,
+        )
+        self._check_event_ts(event_id)
+
+    def test_send_event_ts(self):
+        event_id = self.helper.send(
+            self.room,
+            tok=self.appservice.token,
+            ts=self.ts,
+            appservice_user_id=self.appservice_user,
+        )["event_id"]
+        self._check_event_ts(event_id)
+
+    def test_send_event_msc3316_ts(self):
+        event_id = self.helper.send(
+            self.room,
+            tok=self.appservice.token,
+            msc3316_ts=self.ts,
+            appservice_user_id=self.appservice_user,
+        )["event_id"]
+        self._check_event_ts(event_id)
+
+    def _check_event_ts(self, event_id):
+        res = self.get_success(self.hs.get_datastore().get_event(event_id))
+        self.assertEquals(self.ts, res.origin_server_ts)
```

Comment:
This could do with a comment:

```suggestion
        # check that the event was successfully persisted to the database with the correct timestamp.
        res = self.get_success(self.hs.get_datastore().get_event(event_id))
        self.assertEquals(self.ts, res.origin_server_ts)
```